File: C#/Easy/ValidPalindrome.cs
public class Solution {
    public bool IsPalindrome(string s) {
        int left = 0;
        int right = s.Length - 1;

        while (left <= right)
        {
            while (!Char.IsLetterOrDigit(s[left]) && left < right) left++;
            while (!Char.IsLetterOrDigit(s[right]) && left < right) right--;

            char c1 = s[left];
            char c2 = s[right];
            bool isAlphanumeric = Char.IsLetterOrDigit(c1) && Char.IsLetterOrDigit(c2);
            if (isAlphanumeric && Char.ToLower(c1) != Char.ToLower(c2)) return false;
            
            right--;
            left++;
        }

        return true;
    }
}

/*

  Two pointers on each end ignoring case and non-alphanumerics

  Time Complexity: O(n)
  Space Complexity: O(1)

*/

File: Swift/Easy/ValidPalindrome.swift
class Solution {
    func isPalindrome(_ s: String) -> Bool {
        let s = Array(s.lowercased().filter { $0.isLetter || $0.isNumber })
        var left = 0
        var right = s.count - 1
        while left < right {
            if s[left] != s[right] { return false }

            left += 1
            right -= 1
        }

        return true;
    }
}

/*

    Time: O(n)
    Space: O(n)

*/
